OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+  return slice;
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnContextSwitchOut(
+    uint64_t timestamp_ns, pid_t tid, ThreadStateSlice::ThreadState new_state) {
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_switch(out) but previous state of thread %d is unknown", tid);
+    tid_open_states_.insert_or_assign(tid, OpenState{new_state, timestamp_ns});
+    return std::nullopt;
+  }
+
+  const OpenState& open_state = open_state_it->second;
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    tid_open_states_.insert_or_assign(tid, OpenState{new_state, timestamp_ns});
+    return std::nullopt;
+  }
+
+  // As we are switching out of a CPU, if the previous state was kRunnable, assume it was kRunning.
+  // This is because when we retrieve the initial thread states we have no way to distinguish
+  // between kRunnable and kRunning. After all, "running" is not a real thread state.
+  ThreadStateSlice::ThreadState adjusted_open_state_state = open_state.state;
+  if (adjusted_open_state_state == ThreadStateSlice::kRunnable) {
+    adjusted_open_state_state = ThreadStateSlice::kRunning;
+  }
+
+  if (adjusted_open_state_state != ThreadStateSlice::kRunning) {
+    ERROR("Processed sched:sched_switch(out) for thread %d but unexpected previous state %s", tid,
+          ThreadStateSlice::ThreadState_Name(adjusted_open_state_state));
+    if (adjusted_open_state_state == new_state) {
+      // No state change: do nothing and don't overwrite the previous begin timestamp.
+      return std::nullopt;
+    }
+  }
+
+  ThreadStateSlice slice;
+  slice.set_tid(tid);
+  slice.set_thread_state(adjusted_open_state_state);
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+
+  // Note: If the thread exits but the new_state is kZombie instead of kDead,
+  // the switch to kDead will never be reported.
+  tid_open_states_.insert_or_assign(tid, OpenState{new_state, timestamp_ns});
+  return slice;
+}
+
+std::vector<ThreadStateSlice> ThreadStateManager::OnCaptureFinished(uint64_t timestamp_ns) {
+  std::vector<ThreadStateSlice> slices;
+  for (const auto& [tid, open_state] : tid_open_states_) {
+    ThreadStateSlice slice;
+    slice.set_tid(tid);
+    slice.set_thread_state(open_state.state);
+    slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+    slice.set_end_timestamp_ns(timestamp_ns);
+    slices.emplace_back(std::move(slice));
+  }
+  return slices;
+}
+
+void ThreadStateVisitor::ProcessInitialState(uint64_t timestamp_ns, pid_t tid, char state_char) {
+  std::optional<ThreadStateSlice::ThreadState> initial_state = GetThreadStateFromChar(state_char);
+  if (!initial_state.has_value()) {
```

Comment:
That should be a CHECK, shouldn't it?

Reply:
I understand your point. Especially, as the user experience of a crashing service, is not good atm.
But I still have the impression, that we rely on certain behavior of the system and might be doing something wrong, if that system behaves differently. We probably need the metrics uploader or missuse the crash server, for such kind of errors. Were we notify our system, but do not crash the service.